### ISSUE TITLE
Bazel crashes when specifying target_compatible_with to rust_binary

### ISSUE THREAD

**Issue by duarten** [bazelbuild/bazel#13250]
### Description of the problem / feature request:

When using `rules_rust` and specifying `target_compatible_with`, Bazel crashes:

```
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'ConfiguredTargetKey{label=//:hello, config=BuildConfigurationValue.Key[86e57527e1c8bb10edc853e734cb858f8159d8f3e0a4df9ceb16f80aad784b93]}' (requested by nodes )
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:563)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:398)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.NullPointerException
	at com.google.devtools.build.lib.analysis.platform.ConstraintCollection.hasConstraintValue(ConstraintCollection.java:216)
	at com.google.devtools.build.lib.analysis.RuleContext.targetPlatformHasConstraint(RuleContext.java:1226)
	at com.google.devtools.build.lib.analysis.constraints.RuleContextConstraintSemantics.lambda$incompatibleConfiguredTarget$1(RuleContextConstraintSemantics.java:904)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(Unknown Source)
	at com.google.common.collect.CollectSpliterators$1.lambda$forEachRemaining$1(CollectSpliterators.java:120)
	at java.base/java.util.AbstractList$RandomAccessSpliterator.forEachRemaining(Unknown Source)
	at com.google.common.collect.CollectSpliterators$1.forEachRemaining(CollectSpliterators.java:120)
	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)
	at com.google.devtools.build.lib.analysis.constraints.RuleContextConstraintSemantics.incompatibleConfiguredTarget(RuleContextConstraintSemantics.java:905)
	at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createRule(ConfiguredTargetFactory.java:327)
	at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createConfiguredTarget(ConfiguredTargetFactory.java:194)
	at com.google.devtools.build.lib.skyframe.SkyframeBuildView.createConfiguredTarget(SkyframeBuildView.java:938)
	at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.createConfiguredTarget(ConfiguredTargetFunction.java:1013)
	at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:371)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:477)
	... 7 more
```

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

My WORKSPACE file is:

```starlark
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "rules_rust",
    sha256 = "e6d835ee673f388aa5b62dc23d82db8fc76497e93fa47d8a4afe97abaf09b10d",
    strip_prefix = "rules_rust-f37b9d6a552e9412285e627f30cb124e709f4f7a",
    urls = [
        # Master branch as of 2021-01-27
        "https://github.com/bazelbuild/rules_rust/archive/f37b9d6a552e9412285e627f30cb124e709f4f7a.tar.gz",
    ],
)

load("@rules_rust//rust:repositories.bzl", "rust_repositories")

rust_repositories(version = "1.50.0", edition="2018")
```

And my target is:

```starlark
package(default_visibility = ["//visibility:public"])

load(
    "@rules_rust//rust:rust.bzl",
    "rust_binary",
)

platform(
    name = "linux",
    constraint_values = [
        "@platforms//os:linux",
        "@platforms//cpu:x86_64",
    ],
)

rust_binary(
    name = "hello",
    srcs = ["main.rs"],
    target_compatible_with = ["@rules_rust//rust/platform:linux" ],
)
```

### What operating system are you running Bazel on?

macOS

### What's the output of `bazel info release`?

release 4.0.0

**Comment by philsc** [comment]
Oh dear. That's no good. I'll take a look.

**Comment by philsc** [comment]
Okay, I figured it out. The problem is that `@rules_rust//rust/platform:linux` is a `config_setting()` that wraps the `@platforms//os:linux` constraint.

AFAICT using the `@platforms//os:linux` constraint directly should work for you. `target_compatible_with` doesn't support config settings.

That being said, that crash should instead convey a useful error message. I'll try to get that fixed soon.

Thanks for flagging this!

**Comment by philsc** [comment]
@gregestren FYI.

**Comment by duarten** [comment]
Awesome, thanks for looking into this!

**Comment by philsc** [comment]
Okay, this one took me a bit to untangle.  It turns out the original patch discarded `RuleContext` errors. I submitted a patch that should fix that ( #13254). AFAICT it still passes all the tests.
